### PR TITLE
feat: normalize urls when breaking down

### LINKED ISSUE #12770: Add option to normalize URLs in breakdown
## Is your feature request related to a problem?

Currently URLs "https://example.com/foo" and "https://example.com/foo/" are treated as completely separate entities when breaking down by Current URL. In reality they are likely the same thing though (though not always).

## Describe the solution you'd like

When breaking down by Current URL or Path Name, there should be an option to normalize those paths on the query level, so that `/foo`, `/foo/`, `/foo#` and `/foo?` are treated as equivalent. 

## Additional context

This has been [reported by a customer](https://posthog.slack.com/archives/C021HLFPAA2/p1668112961895649).

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

In [this slack thread a customer reported](https://posthog.slack.com/archives/C045L1VEG87/p1668506350569719) that when breaking down by "Current URL" or "Path" they are seeing unexpected results

If they have events with paths like `/home/` and `/home` they are being counted as different URLs

closes #12770

## Changes

* Adds a new breakdown filter option to normalize URLs
* Extends `BreakdownTag` to let people set it
* when set trims trailing `/` or `?` or `#` from current url or path
* default to normalizing URL and pathname

### UI

Screenshot 2022-11-15 at 23 05 46
Screenshot 2022-11-15 at 23 05 51

### with setting on

Screenshot 2022-11-15 at 23 01 08

### with setting off

Screenshot 2022-11-15 at 23 01 31

### by current URL

breakdown-by-current-url

### by pathname

breakdown-by-pathname

## How did you test this code?

* added developer tests
* ran it locally

### CODE REVIEW

File: frontend/src/scenes/insights/filters/BreakdownFilter/TaxonomicBreakdownFilter.tsx [hunk 1]
```diff
@@ -61,7 +62,13 @@ export function BreakdownFilter({
                           } else {
                               const newParts = breakdownParts.filter((_, i) => i !== index)
                               setFilters({
-                                  breakdowns: newParts.map((np): Breakdown => ({ property: np, type: breakdown_type })),
+                                  breakdowns: newParts.map(
+                                      (np): Breakdown => ({
+                                          property: np,
+                                          type: breakdown_type,
+                                          normalize_url: isURLNormalizeable(np.toString()),
```
Comment (mariusandra) [inline frontend/src/scenes/insights/filters/BreakdownFilter/TaxonomicBreakdownFilter.tsx:69]: Will we send this property over with other non-url breakdown params as well? I'd prefer then to not send the key at all.

File: frontend/src/scenes/insights/filters/BreakdownFilter/index.ts [hunk 2]
```diff
@@ -1 +1,5 @@
 export { BreakdownFilter } from './TaxonomicBreakdownFilter'
+
+export const isURLNormalizeable = (propertyName: string): boolean => {
+    return ['$current_url', '$pathname'].includes(propertyName)
```
Comment (mariusandra) [inline frontend/src/scenes/insights/filters/BreakdownFilter/index.ts:4]: perhaps also `propertyName.endsWith('_url')`?

**Review by mariusandra — APPROVED** [review]
Some nits, but looks great! Thanks for jumping on this so quickly!

File: frontend/src/scenes/insights/filters/BreakdownFilter/index.ts (hunk 2 above)
Comment (pauldambra) [reply frontend/src/scenes/insights/filters/BreakdownFilter/index.ts:4]: I considered that but I think the next step would be to add a property type of URL and _maybe_ autodetect that

File: frontend/src/scenes/insights/filters/BreakdownFilter/TaxonomicBreakdownFilter.tsx (hunk 1 above)
Comment (pauldambra) [reply frontend/src/scenes/insights/filters/BreakdownFilter/TaxonomicBreakdownFilter.tsx:69]: Yep, https://github.com/PostHog/posthog/pull/12808/files#diff-5acf0d9fad9469e30132e1a2ef62c761a5f2837c1a0da79670b14fb4d6c8af2cR122

File: frontend/src/scenes/insights/filters/BreakdownFilter/BreakdownTag.tsx [hunk 3]
```diff
@@ -1,14 +1,18 @@
...
 import { breakdownTagLogic } from './breakdownTagLogic'
 import { isAllCohort, isCohort, isPersonEventOrGroup } from './TaxonomicBreakdownFilter'
+import { Tooltip } from 'lib/components/Tooltip'
+import { IconInfo } from 'lib/components/icons'
 
 type BreakdownTagProps = {
     isHistogramable: boolean
-    setFilters?: (filter: Partial<FilterType>) => void
+    // e.g. example.com and example.com/ should be treated as the same value
+    isURLNormalizeable: boolean
```
Comment (Twixes) [inline frontend/src/scenes/insights/filters/BreakdownFilter/BreakdownTag.tsx:14]: VERY tiny nit: "URL" is camel-cased in `setNormalizeBreakdownUrl`, but all capitalized here

File: posthog/models/filters/mixins/common.py [hunk 4]
```diff
@@ -240,6 +243,14 @@ def breakdown_type_and_group_to_dict(self):
         else:
             return {}
 
+    @cached_property
+    def breakdown_normalize_url(self) -> bool:
+        """
+        When breaking down by $current_url or $pathname, we ignore trailing slashes and question marks
```
Comment (Twixes) [inline posthog/models/filters/mixins/common.py:249]: Hashes too, right?

File: posthog/queries/trends/breakdown.py [hunk 5]
```diff
@@ -403,7 +403,11 @@ def _get_breakdown_value(self, breakdown: str) -> str:
                 breakdown_value, _ = get_property_string_expr("events", breakdown, "%(key)s", "properties")
 
         if self.filter.using_histogram:
-            return f"toFloat64OrNull(toString({breakdown_value}))"
+            breakdown_value = f"toFloat64OrNull(toString({breakdown_value}))"
+
+        if self.filter.breakdown_normalize_url:
+            breakdown_value = f"if( length(trim(TRAILING '/?#' from {breakdown_value})) = 0, '/', trim(TRAILING '/?#' from {breakdown_value}))"
```
Comment (Twixes) [inline posthog/queries/trends/breakdown.py:409]: Clickhouse also has an `empty()` function, might be a bit more readable. Also I suppose, we could do the `trim(TRAILING)` operation once if it was defined as an extra column (`... AS raw_normalized_breakdown_value, if(...)`)

**Review by Twixes — COMMENTED** [review]
Thanks so much for taking this on, I love this

I have some UI suggestions, but basically I just put them into #12813, since it felt silly to edit the frontend code to show what I mean and then _not_ actually propose that code (and going into Figma would be a huge time sink).

File: posthog/queries/trends/breakdown.py (hunk 5 above)
Comment (Twixes) [reply posthog/queries/trends/breakdown.py]: Ah yeah, I didn't take a good look at the complexity of that. It should be fine as is too.

**Pull request merged**